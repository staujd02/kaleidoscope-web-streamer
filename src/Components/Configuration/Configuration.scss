@import '../../Assets/styling/mixins.scss';
@import '../../Assets/styling/palette.scss';

.title {
    border-bottom: 1px $color4 solid;
}

.config-container {
    padding-top: 38px;
}

div.source-list {
    ul {
        border: 1px rgba($color1, $heavy) solid;
        max-height: 190px;
        overflow: scroll;
        padding-bottom: 0;
        padding-top: 0;

        .Mui-selected {
            color: rgba($color4, $heavy);
            background-color: rgba($color1, $heavy);

            &:hover {
                color: rgba($color1, $heavy);
            }
        }

        &>div {
            color: rgba($color1, $heavy);

            &:active {
                color: rgba($color3, $heavy);
            }

            &:hover {
                color: rgba($color4, $heavy);
                background-color: rgba($color1, $heavy);
            }
        }
    }
}

button.resume-cycle {
    @include alpha-background-color(rgba($color1, $heavy), $color2);

    position: absolute;
    top: 8px;
    background-color: $color2;
    color: $color4;

    &:hover {
        background-color: $color4;
        color: $color2;
        border: 1px solid $color2;
    }
}

.card {
    flex-grow: 1;
}

.placeholder {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 75%;
    font-style: italic;
}

.search {
    max-height: 40px;
    width: 100%;
}