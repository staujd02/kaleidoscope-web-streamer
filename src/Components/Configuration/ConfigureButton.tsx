import React from 'react';
import { Button } from '@material-ui/core';

export default class ConfigureButton extends React.Component<ConfigureButtonProps> {
    render(){
        const { buttonClickHandler } = this.props;
        return <Button color="primary" className="configure-button" onClick={buttonClickHandler}>Configure Streams</Button>
    }
}