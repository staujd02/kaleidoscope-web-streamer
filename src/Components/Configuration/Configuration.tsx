import React from 'react';
import './Configuration.scss';
import SourceDetails from './SourceDetails/SourceDetails';
import ConfigurationHeader from '../Headers/ConfigurationHeader';
import SourceListCard from './SourceListCard/SourceListCard';
import StreamPreview from './StreamPreview/StreamPreview';
import Button from '@material-ui/core/Button';
import Container from '@material-ui/core/Container';
import SourceRepositoryMutator from '../Mutators/SourceRepositoryMutator';

export default class Configuration extends React.Component<ConfigurationProps, ConfigurationState> {

    constructor(props: ConfigurationProps) {
        super(props);
        this.state = {
            selectedSource: null
        }
    }

    handleSelectSource = (s: Source) => {
        this.setState({
            selectedSource: s.key
        });
    }
    
    sortDown = () => {
        this.props.handleSave(
            SourceRepositoryMutator.sortDown(this.props.sourceRepo, this.state.selectedSource)
        )
    }

    sortUp = () => {
        this.props.handleSave(
            SourceRepositoryMutator.sortUp(this.props.sourceRepo, this.state.selectedSource)
        )
    }

    addSource = () => {
        let s;
        this.props.handleSave(
            (s = SourceRepositoryMutator.add(this.props.sourceRepo))
        );
        this.setState({ selectedSource: s.streams[s.streams.length - 1].key });
    }

    deleteSelectedSource = () => {
        this.props.handleSave(
            SourceRepositoryMutator.delete(this.props.sourceRepo, this.state.selectedSource)
        );
        this.setState({ selectedSource: null });
    }

    handleUpdate = (s: Source) =>
        this.props.handleSave(
            SourceRepositoryMutator.update(this.props.sourceRepo, s)
        );

    render() {
        let source = this.props.sourceRepo.streams.find(s => s.key === this.state.selectedSource) || null;
        const sortUpDisabled = source === null ? true : !this.props.sourceRepo.streams.find(s => s.sortOrder < source!.sortOrder);
        const sortDownDisabled = source === null ? true : !this.props.sourceRepo.streams.find(s => s.sortOrder > source!.sortOrder);
        return (
            <ConfigurationHeader>
                <Container className="config-titlebar">
                    <Button onClick={this.props.doneConfiguring} className="resume-cycle">Resume Cycle</Button>
                </Container>
                <Container className="config-content">
                    <SourceListCard
                        handleSelectSource={this.handleSelectSource}
                        sortUp={this.sortUp}
                        sortDown={this.sortDown}
                        sortDisabled={[sortUpDisabled, sortDownDisabled]}
                        addSource={this.addSource}
                        deleteSource={this.deleteSelectedSource}
                        sourceList={this.props.sourceRepo.streams} />
                    <SourceDetails handleUpdate={this.handleUpdate} source={source} />
                    <StreamPreview source={source} />
                </Container>
            </ConfigurationHeader>
        );
    }

}