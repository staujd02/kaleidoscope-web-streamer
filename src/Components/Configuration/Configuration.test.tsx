import React from 'react';
import { formatHTML } from '../../TestUtilities/htmlFormatter';
import { shallow, ShallowWrapper } from 'enzyme';
import Configuration from './Configuration';

describe('The Configuration Component', () => {

    let wrapper: ShallowWrapper<ConfigurationProps, ConfigurationState, Configuration>
    let sourceRepository: SourceRepository
    let handleSave: jest.Mock;
    let doneConfiguring: jest.Mock;
    let props: ConfigurationProps;

    beforeEach(() => {
        props = {
            sourceRepo: (sourceRepository = setupRepo(sourceRepository)),
            handleSave: (handleSave = jest.fn()),
            doneConfiguring: (doneConfiguring = jest.fn())
        }
        wrapper = shallow(<Configuration {...props} />);
    });

    it('renders correctly', () => expect(formatHTML(wrapper.html())).toMatchSnapshot());

    it('renders the list of sources through the source list', () => {
        expect(wrapper.find('SourceListCard').prop('sourceList')).toEqual(sourceRepository.streams);
    });

    it('renders the currently selected source ', () => {
        expect(wrapper.find('SourceDetails').prop('source')).toEqual(wrapper.state().selectedSource);
    });

    it('renders a button to restart the streams', () => {
        (wrapper.find('.resume-cycle').prop('onClick') as () => void)();
        expect(doneConfiguring).toHaveBeenCalled();
    });

    describe('when a user updates a source', () => {

        let updatedSource: Source;

        beforeEach(() => {
            updatedSource = sourceRepository.streams[2];
            wrapper.instance().handleSelectSource(updatedSource);
            updatedSource.title = "New Title";
            (wrapper
                .find('SourceDetails')
                .prop('handleUpdate') as (s: Source) => void)
                (updatedSource);
        })

        it('saves the changes to that source and that source alone', () => {
            sourceRepository.streams[2] = updatedSource;
            expect(handleSave).toHaveBeenCalledWith(sourceRepository);
        });
    });

    describe('when the user selects a source', () => {

        beforeEach(() => {
            (wrapper
                .find('SourceListCard')
                .prop('handleSelectSource') as (s: Source) => void)
                (sourceRepository.streams[0]);
        })

        it('updates the detail page', () => {
            expect(wrapper.find('SourceDetails').prop('source')).toEqual(sourceRepository.streams[0]);
        });

        it('updates the preview stream', () => {
            expect(wrapper.find('StreamPreview').prop('source')).toEqual(sourceRepository.streams[0]);
        });
    });


    function setupRepo(sourceRepository: SourceRepository) {
        sourceRepository = {
            streams: [{
                source: 'httpLink1...',
                title: 'Source 1',
                isEnabled: true,
                duration: 60,
                sortOrder: 1,
                key: 1
            },
            {
                source: 'httpLink3...',
                title: 'Source 3',
                isEnabled: true,
                duration: 60,
                sortOrder: 3,
                key: 4
            },
            {
                source: 'httpLink2..',
                title: 'Source 2',
                isEnabled: true,
                duration: 60,
                sortOrder: 2,
                key: 2
            },
            {
                source: 'httpLink4..',
                title: 'Source 4',
                isEnabled: false,
                duration: 60,
                sortOrder: 4,
                key: 15
            }
            ]
        };
        return sourceRepository;
    }
});
