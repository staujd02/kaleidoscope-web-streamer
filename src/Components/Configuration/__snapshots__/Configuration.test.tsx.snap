// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Configuration Component renders correctly 1`] = `
"
<header class=\\"config-viewport\\">
  <div class=\\"config-container\\">
    <div class=\\"MuiContainer-root config-titlebar MuiContainer-maxWidthLg\\">
      <button class=\\"MuiButtonBase-root MuiButton-root resume-cycle MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\">
        <span class=\\"MuiButton-label\\">Resume Cycle</span>
      </button>
    </div>
    <div class=\\"MuiContainer-root config-content MuiContainer-maxWidthLg\\">
      <div class=\\"MuiPaper-root MuiPaper-elevation1 MuiCard-root card source-list MuiPaper-rounded\\">
        <div class=\\"MuiCardContent-root\\">
          <p class=\\"MuiTypography-root title MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom\\">List of the Streams</p>
          <div class=\\"MuiFormControl-root MuiTextField-root search MuiFormControl-marginDense\\">
            <label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined\\" data-shrink=\\"false\\" for=\\"outlined-dense\\">Search</label>
            <div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense\\">
              <input type=\\"text\\" aria-invalid=\\"false\\" class=\\"MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense\\" id=\\"outlined-dense\\">
              <fieldset aria-hidden=\\"true\\" style=\\"padding-left:8px\\" class=\\"PrivateNotchedOutline-root-144 MuiOutlinedInput-notchedOutline\\">
                <legend class=\\"PrivateNotchedOutline-legend-145\\" style=\\"width:0.01px\\">
                  <span>â€‹</span>
                </legend>
              </fieldset>
            </div>
          </div>
          <ul class=\\"MuiList-root MuiList-padding\\" aria-label=\\"streams\\">
            <div class=\\"MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">
              <div class=\\"MuiListItemText-root\\">
                <span class=\\"MuiTypography-root MuiListItemText-primary MuiTypography-body1\\">Source 1</span>
              </div>
            </div>
            <div class=\\"MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">
              <div class=\\"MuiListItemText-root\\">
                <span class=\\"MuiTypography-root MuiListItemText-primary MuiTypography-body1\\">Source 2</span>
              </div>
            </div>
            <div class=\\"MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">
              <div class=\\"MuiListItemText-root\\">
                <span class=\\"MuiTypography-root MuiListItemText-primary MuiTypography-body1\\">Source 3</span>
              </div>
            </div>
            <div class=\\"MuiButtonBase-root MuiListItem-root disabled MuiListItem-gutters MuiListItem-button\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\">
              <div class=\\"MuiListItemText-root\\">
                <span class=\\"MuiTypography-root MuiListItemText-primary MuiTypography-body1\\">Source 4</span>
              </div>
            </div>
          </ul>
        </div>
      </div>
      <div class=\\"MuiPaper-root MuiPaper-elevation1 MuiCard-root card source-detail MuiPaper-rounded\\">
        <div class=\\"MuiCardContent-root\\">
          <p class=\\"MuiTypography-root title MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom\\">Stream Details</p>
          <p class=\\"MuiTypography-root placeholder MuiTypography-body1\\">Select a stream to view it's details.</p>
        </div>
      </div>
      <div class=\\"MuiPaper-root MuiPaper-elevation1 MuiCard-root card stream-preview MuiPaper-rounded\\">
        <div class=\\"MuiCardContent-root\\">
          <p class=\\"MuiTypography-root title MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom\\">Preview of the Stream</p>
          <p class=\\"MuiTypography-root placeholder MuiTypography-body1\\">Select a stream to view a preview of it.</p>
        </div>
      </div>
    </div>
  </div>
</header>
"
`;
